### ISSUE TITLE
Explicit Caching of Images from Mounted Filesystems

### ISSUE THREAD

**Issue by tri-adam** [apptainer/singularity#2814]
Consider a scenario where we have a cluster with a parallel filesystem, where each node also has some sort of (faster) local storage. In this scenario, the workflow includes running the same image more than once on many nodes, and the image is present on the parallel filesystem. We would like to use the local storage to speed up repeated access to images, and reduce the load on the parallel filesystem. 

The idea would be to use the local storage as a cache to speed up repeated accesses to images. The first time a particular image is run, it would be pulled from the parallel filesystem, and cached on the local storage of the node. Subsequent runs of the image should leverage the local cache of the image (presuming the cache hasn't been cleared.)

This behaviour occurs already if the image is pulled from an external source. For example:

```sh
adam@ubuntu:~$ singularity shell library://alpine
INFO:    Downloading library image
 2.08 MiB / 2.08 MiB [==========================================================] 100.00% 5.94 MiB/s 0s
Singularity> exit
adam@ubuntu:~$ singularity shell library://alpine
Singularity> exit
```

One idea would be to use a URI to explicitly enable caching for file access:

```sh
singularity shell filecache://<path-to-parallel-fs>
```

Or maybe a flag could be added that allows a user to explicitly enable caching when specifying a file:

```sh
singularity shell --use-cache <path-to-parallel-fs>
```

**Comment by truatpasteurdotfr** [comment]
how is it different from testing if the local sif file is already there? if not, then copy it from the parallel filesystem (if there), or pull it from upstream? 

something like:
```
#!/bin/sh
sif=alpine
[ -f /parallel/cache/${sif} ] || singularity pull /parallel/cache/${sif}  library://${sif}
[ -f /local/scratch/${sif| ] || cp /parallel/cache/${sif} /local/scratch/${sif}
singularity shell /local/scratch/${sif}
```
that quick and dirty script does not adress the version changes of the container :P

**Comment by GodloveD** [comment]
I agree with @truatpasteurdotfr that this is something that user can likely already do if they need to just with scripting.  And it actually might prove to be very confusing to users if this is carried out automatically on their behalf.  So, I'm not really sure it would be an improvement overall.

**Comment by mem** [comment]
While I understand the original request (I can rephrase it as "force the use of the cache for local:// images") the way the cache is implemented actually makes it a little cumbersome to implement.

The cache is implemented with registries that provide some kind of hashing in mind. When you run `singularity pull library://alpine:3.9` it will first request the image's hash and use that to find the corresponding image in the cache. Since the implementation of the cache is different for each type of provider, there's nothing that says that something like `singularity pull alpine_3.9.sif /path/to/alpine:3.9` shouldn't work (it could for example stat the file and make sure everything matches; the problem is that "everything" would have to be the size and nothing else because it's reasonable to expect that ownership and timestamps are different). If we were to read the remote file to compute a hash, we might copy the file as well (we would be saving one write but we would have to read twice, and for large enough images, where this makes sense, both reads would have to actually happen).

Short version: while I agree with the basic premise, I doubt there's an actual benefit to implementing this.

You are probably better off with a local caching proxy (this happens to be the second time today that I have said this).

**Comment by mem** [comment]
And just for the record: this is not just a straight http caching proxy, as most of the traffic goes over https, so it's not possible to actually cache this (the connection is encrypted, so a normal proxy does not get to see the actual URLs being requested, as the client will request a tunnel using the CONNECT method).
